cmd/juju/cloud/add.go

```diff
@@ -302,29 +299,70 @@ func (c *AddCloudCommand) Run(ctxt *cmd.Context) error {
 	if c.Replace {
 		ctxt.Warningf("'add-cloud --replace' is deprecated. Use 'update-cloud' instead.")
 	}
-	if c.CloudFile == "" && c.controllerName == "" {
-		return c.runInteractive(ctxt)
-	}
-
 	var newCloud *jujucloud.Cloud
 	var err error
 	if c.CloudFile != "" {
 		newCloud, err = c.readCloudFromFile(ctxt)
 	} else {
-		// No cloud file specified so we try and use a named
-		// cloud that already has been added to the local cache.
-		newCloud, err = cloudFromLocal(c.Store, c.Cloud)
+		if c.Cloud != "" {
+			// It's possible that the user wants to add an existing local client to a current controller,
+			// i.e. 'juju add-cloud aws'. So let's see if we can find the cloud.
+			newCloud, err = common.CloudByName(c.Cloud)
+			if err == nil {
+				c.existsLocally = newCloud.Name != ""
+			}
+		}
+		if !c.existsLocally {
+			newCloud, err = c.runInteractive(ctxt)
+		}
 	}
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	if c.controllerName == "" {
-		if !c.Local {
-			ctxt.Infof(
-				"There are no controllers running.\nAdding cloud to local cache so you can use it to bootstrap a controller.\n")
+	// All clouds must have at least one default region - lp#1819409.
+	if len(newCloud.Regions) == 0 {
+		newCloud.Regions = []jujucloud.Region{{Name: "default"}}
```

Comment:
It may need to be done in both location.

The client needs to have it since we storing this cloud locally. The server side will need it for all other clients like pyjuju. I do not want to do this here as I do not like big unfocused PRs that take for ever to land. I'll follow up.